# **Market Basket Analysis with Apriori Algorithm**

![image](../images/cover-image.png)

<hr>

**`In this Notebook we would learn`** 

1. What is Association Rule Learning?

2. How Apriori works ?

3. Implementing Apriori With Python - in 2 ways

**Problem Statement :**

When we go grocery shopping, we often have a standard list of things to buy. Each shopper has a distinctive list, depending on one’s needs and preferences. A housewife might buy healthy ingredients for a family dinner, while a bachelor might buy beer and chips. Understanding these buying patterns can help to increase sales in several ways. If there is a pair of items, X and Y, that are frequently bought together:

> Both X and Y can be placed on the same shelf, so that buyers of one item would be prompted to buy the other.

> Promotional discounts could be applied to just one out of the two items.

> Advertisements on X could be targeted at buyers who purchase Y.

> X and Y could be combined into a new product, such as having Y in flavors of X.

> **While we may know that certain items are frequently bought together, the question is, how do we uncover these associations?**

*Besides increasing sales profits, association rules can also be used in other fields. In medical diagnosis for instance, understanding which symptoms tend to co-morbid can help to improve patient care and medicine prescription.*


What is Association Rule Learning?
--

Association Rule Learning is rule-based learning for identifying the association between different variables in a database. **`One of the best and most popular examples of Association Rule Learning is the Market Basket Analysis`**. The problem analyses the association between various items that has the highest probability of being bought together by a customer.

For example, the association rule, {onions, chicken masala} => {chicken} says that a person who has got both onions and chicken masala in his or her basket has a high probability of buying chicken also.


Apriori Algorithm
--

The algorithm was first proposed in 1994 by Rakesh Agrawal and Ramakrishnan Srikant. Apriori algorithm finds the most frequent itemsets or elements in a transaction database and identifies association rules between the items just like the above-mentioned example.


How Apriori works ?
--

To construct association rules between elements or items, the algorithm considers 3 important factors which are, support, confidence and lift. Each of these factors is explained as follows:

**Support:**

The support of item I is defined as the ratio between the number of transactions containing the item I by the total number of transactions expressed as :

<img src="https://miro.medium.com/max/403/0*pyOADkeaWyrVP2ft.png" />

<font color='darkgreen'><b>Support</b></font> indicates how popular an itemset is, as measured by the proportion of transactions in which an itemset appears. In Table 1 below, the support of {apple} is 4 out of 8, or 50%. Itemsets can also contain multiple items. For instance, the support of {apple, beer, rice} is 2 out of 8, or 25%.

<img src='https://annalyzin.files.wordpress.com/2016/04/association-rule-support-table.png?w=503&h=447' />

<br />

<font color='red'><em>If you discover that sales of items beyond a certain proportion tend to have a significant impact on your profits, you might consider using that proportion as your support threshold. You may then identify itemsets with support values above this threshold as significant itemsets.</em></font>


**Confidence:**

This is measured by the proportion of transactions with item I1, in which item I2 also appears. The confidence between two items I1 and I2,  in a transaction is defined as the total number of transactions containing both items I1 and I2 divided by the total number of transactions containing I1. ( Assume I1 as X , I2 as Y )

<img src='https://miro.medium.com/max/576/1*50GI4dR58MnhwBP9dw6nFQ.png' />

<font color='darkgreen'><b>Confidence</b></font> says how likely item Y is purchased when item X is purchased, expressed as {X -> Y}. This is measured by the proportion of transactions with item X, in which item Y also appears. In Table 1, the confidence of {apple -> beer} is 3 out of 4, or 75%.

<img src='https://annalyzin.files.wordpress.com/2016/03/association-rule-confidence-eqn.png?w=527&h=77' />

<font color='red'><em>One drawback of the confidence measure is that it might misrepresent the importance of an association. This is because it only accounts for how popular apples are, but not beers. If beers are also very popular in general, there will be a higher chance that a transaction containing apples will also contain beers, thus inflating the confidence measure. To account for the base popularity of both constituent items, we use a third measure called lift. </em></font>


**Lift:**

Lift is the ratio between the confidence and support.

<font color='darkgreen'><b>Lift</b></font> says how likely item Y is purchased when item X is purchased, while controlling for how popular item Y is. In Table 1, the lift of {apple -> beer} is 1,which implies no association between items. A lift value greater than 1 means that item Y is likely to be bought if item X is bought, while a value less than 1 means that item Y is unlikely to be bought if item X is bought. ( *here X represents apple and Y represents beer* )

<img src='https://annalyzin.files.wordpress.com/2016/03/association-rule-lift-eqn.png?w=566&h=80' />

<hr>

for **Extra Reading** : refer <a href='https://towardsdatascience.com/association-rules-2-aa9a77241654'> this </a> 


Implementing Apriori With Python - in 2 ways 
--

<h3>In the <b><u>first way</u></b> , we use apyori package.</h3>


In [12]:
#External package need to install
!pip install apyori

Defaulting to user installation because normal site-packages is not writeable


In [13]:
#import all required packages..
import pandas as pd
import numpy as np
from apyori import apriori

In [15]:
#loading market basket dataset..

df = pd.read_csv('../data/Market_Basket_Optimisation.csv',header=None)

In [16]:
df.head()

0          1           2                 3             4   \
0         shrimp    almonds     avocado    vegetables mix  green grapes   
1        burgers  meatballs        eggs               NaN           NaN   
2        chutney        NaN         NaN               NaN           NaN   
3         turkey    avocado         NaN               NaN           NaN   
4  mineral water       milk  energy bar  whole wheat rice     green tea   

                 5     6               7             8             9   \
0  whole weat flour  yams  cottage cheese  energy drink  tomato juice   
1               NaN   NaN             NaN           NaN           NaN   
2               NaN   NaN             NaN           NaN           NaN   
3               NaN   NaN             NaN           NaN           NaN   
4               NaN   NaN             NaN           NaN           NaN   

               10         11     12     13             14      15  \
0  low fat yogurt  green tea  honey  salad  mineral water  salmon   
1             NaN        NaN    NaN    NaN            NaN     NaN   
2             NaN        NaN    NaN    NaN            NaN     NaN   
3             NaN        NaN    NaN    NaN            NaN     NaN   
4             NaN        NaN    NaN    NaN            NaN     NaN   

                  16               17       18         19  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1                NaN              NaN      NaN        NaN  
2                NaN              NaN      NaN        NaN  
3                NaN              NaN      NaN        NaN  
4                NaN              NaN      NaN        NaN

In [17]:
## Data Cleaning step

# replacing empty value with 0.
df.fillna(0,inplace=True)

In [18]:
df.head()

0          1           2                 3             4   \
0         shrimp    almonds     avocado    vegetables mix  green grapes   
1        burgers  meatballs        eggs                 0             0   
2        chutney          0           0                 0             0   
3         turkey    avocado           0                 0             0   
4  mineral water       milk  energy bar  whole wheat rice     green tea   

                 5     6               7             8             9   \
0  whole weat flour  yams  cottage cheese  energy drink  tomato juice   
1                 0     0               0             0             0   
2                 0     0               0             0             0   
3                 0     0               0             0             0   
4                 0     0               0             0             0   

               10         11     12     13             14      15  \
0  low fat yogurt  green tea  honey  salad  mineral water  salmon   
1               0          0      0      0              0       0   
2               0          0      0      0              0       0   
3               0          0      0      0              0       0   
4               0          0      0      0              0       0   

                  16               17       18         19  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1                  0                0        0          0  
2                  0                0        0          0  
3                  0                0        0          0  
4                  0                0        0          0

In [19]:
# Data Pre-processing step

# for using aprori , need to convert data in list format..
# transaction = [['apple','almonds'],['apple'],['banana','apple']]....

transactions = []

for i in range(0,len(df)):
    transactions.append([str(df.values[i,j]) for j in range(0,20) if str(df.values[i,j])!='0'])

In [20]:
## verifying - by printing the 0th transaction
transactions[0]

['shrimp',
 'almonds',
 'avocado',
 'vegetables mix',
 'green grapes',
 'whole weat flour',
 'yams',
 'cottage cheese',
 'energy drink',
 'tomato juice',
 'low fat yogurt',
 'green tea',
 'honey',
 'salad',
 'mineral water',
 'salmon',
 'antioxydant juice',
 'frozen smoothie',
 'spinach',
 'olive oil']

In [21]:
## verifying - by printing the 1st transaction
transactions[1]

['burgers', 'meatballs', 'eggs']

In [23]:
# Call apriori function which requires minimum support, confidance and lift, min length is combination 
# of item default is 2".
rules = apriori(transactions, min_support=0.003, min_confidance=0.2, min_lift=3, min_length=2)

## min_support = 0.003 -> means selecting items with min support of 0.3%
## min_confidance = 0.2 -> means min confidance of 20% 
## min_lift = 3  
## min_length = 2 -> means no. of items in the transaction should be 2

In [24]:
#it generates a set of rules in a generator file...
rules

<generator object apriori at 0x7f214225be60>

In [25]:
# all rules need to be converted in a list..
Results = list(rules)
Results

[RelationRecord(items=frozenset({'brownies', 'cottage cheese'}), support=0.0034662045060658577, ordered_statistics=[OrderedStatistic(items_base=frozenset({'brownies'}), items_add=frozenset({'cottage cheese'}), confidence=0.10276679841897232, lift=3.225329518580382), OrderedStatistic(items_base=frozenset({'cottage cheese'}), items_add=frozenset({'brownies'}), confidence=0.10878661087866107, lift=3.2253295185803816)]),
 RelationRecord(items=frozenset({'light cream', 'chicken'}), support=0.004532728969470737, ordered_statistics=[OrderedStatistic(items_base=frozenset({'chicken'}), items_add=frozenset({'light cream'}), confidence=0.07555555555555556, lift=4.843950617283951), OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)]),
 RelationRecord(items=frozenset({'escalope', 'mushroom cream sauce'}), support=0.005732568990801226, ordered_statistics=[OrderedStatistic(items_base=frozenset({'escalope'}),

In [26]:
# convert result in a dataframe for further operation...
df_results = pd.DataFrame(Results)

In [27]:
# as we see "order_statistics" , is itself a list so need to be converted in proper format..
df_results.head()

items  support  \
0        (brownies, cottage cheese)    0.003   
1            (light cream, chicken)    0.005   
2  (escalope, mushroom cream sauce)    0.006   
3                 (escalope, pasta)    0.006   
4       (fresh bread, tomato juice)    0.004   

                                  ordered_statistics  
0  [((brownies), (cottage cheese), 0.102766798418...  
1  [((chicken), (light cream), 0.0755555555555555...  
2  [((escalope), (mushroom cream sauce), 0.072268...  
3  [((escalope), (pasta), 0.07394957983193277, 4....  
4  [((fresh bread), (tomato juice), 0.09907120743...

In [28]:
# keep support in a separate data frame so we can use later.. 
support = df_results.support

In [29]:
'''
convert orderstatistic in a proper format.
order statistic has lhs => rhs as well rhs => lhs 
we can choose any one for convience.
Let's choose first one which is 'df_results['ordered_statistics'][i][0]'
''' 

#all four empty list which will contain lhs, rhs, confidance and lift respectively.
first_values = []
second_values = []
third_values = []
fourth_value = []

# loop number of rows time and append 1 by 1 value in a separate list.. 
# first and second element was frozenset which need to be converted in list..
for i in range(df_results.shape[0]):
    single_list = df_results['ordered_statistics'][i][0]
    first_values.append(list(single_list[0]))
    second_values.append(list(single_list[1]))
    third_values.append(single_list[2])
    fourth_value.append(single_list[3])

In [30]:
# convert all four list into dataframe for further operation..
lhs = pd.DataFrame(first_values)
rhs = pd.DataFrame(second_values)

confidance=pd.DataFrame(third_values,columns=['Confidance'])

lift=pd.DataFrame(fourth_value,columns=['lift'])

In [31]:
# concat all list together in a single dataframe
df_final = pd.concat([lhs,rhs,support,confidance,lift], axis=1)
df_final

0              1                     0              1  \
0      brownies           None        cottage cheese           None   
1       chicken           None           light cream           None   
2      escalope           None  mushroom cream sauce           None   
3      escalope           None                 pasta           None   
4   fresh bread           None          tomato juice           None   
..          ...            ...                   ...            ...   
89  ground beef       pancakes             spaghetti  mineral water   
90  ground beef           None             spaghetti       tomatoes   
91    olive oil           None                  milk  mineral water   
92         milk  mineral water             spaghetti         shrimp   
93     tomatoes           None                  milk  mineral water   

                2  support  Confidance  lift  
0            None    0.003       0.103 3.225  
1            None    0.005       0.076 4.844  
2            None    0.006       0.072 3.791  
3            None    0.006       0.074 4.701  
4            None    0.004       0.099 3.259  
..            ...      ...         ...   ...  
89           None    0.003       0.211 3.533  
90  mineral water    0.003       0.031 3.344  
91      spaghetti    0.003       0.051 3.217  
92           None    0.003       0.064 3.014  
93      spaghetti    0.003       0.049 3.098  

[94 rows x 8 columns]

In [33]:
'''
 we have some of place only 1 item in lhs and some place 3 or more so we need to a 
 proper represenation for User to understand. 
 replacing none with ' ' and combining three column's in 1 
 example : coffee,none,none is converted to coffee, ,
'''
df_final.fillna(value=' ', inplace=True)
df_final.head()

0  1                     0  1  2  support  Confidance  lift
0     brownies           cottage cheese          0.003       0.103 3.225
1      chicken              light cream          0.005       0.076 4.844
2     escalope     mushroom cream sauce          0.006       0.072 3.791
3     escalope                    pasta          0.006       0.074 4.701
4  fresh bread             tomato juice          0.004       0.099 3.259

In [34]:
#set column name
df_final.columns = ['lhs',1,'rhs',2,3,'support','confidance','lift']
df_final.head()

lhs  1                   rhs  2  3  support  confidance  lift
0     brownies           cottage cheese          0.003       0.103 3.225
1      chicken              light cream          0.005       0.076 4.844
2     escalope     mushroom cream sauce          0.006       0.072 3.791
3     escalope                    pasta          0.006       0.074 4.701
4  fresh bread             tomato juice          0.004       0.099 3.259

In [35]:
# add all three column to lhs itemset only
df_final['lhs'] = df_final['lhs'] + str(", ") + df_final[1]

df_final['rhs'] = df_final['rhs']+str(", ")+df_final[2] + str(", ") + df_final[3]

In [36]:
df_final.head()

lhs  1                         rhs  2  3  support  confidance  \
0     brownies,             cottage cheese,  ,            0.003       0.103   
1      chicken,                light cream,  ,            0.005       0.076   
2     escalope,       mushroom cream sauce,  ,            0.006       0.072   
3     escalope,                      pasta,  ,            0.006       0.074   
4  fresh bread,               tomato juice,  ,            0.004       0.099   

   lift  
0 3.225  
1 4.844  
2 3.791  
3 4.701  
4 3.259

In [37]:
#drop columns 1,2 and 3 because now we already appended to lhs column.

df_final.drop(columns=[1,2,3],inplace=True)

In [38]:
#this is final output. You can sort based on the support lift and confidance..
df_final.head()

lhs                         rhs  support  confidance  lift
0     brownies,          cottage cheese,  ,      0.003       0.103 3.225
1      chicken,             light cream,  ,      0.005       0.076 4.844
2     escalope,    mushroom cream sauce,  ,      0.006       0.072 3.791
3     escalope,                   pasta,  ,      0.006       0.074 4.701
4  fresh bread,            tomato juice,  ,      0.004       0.099 3.259

In [39]:
## Showing top 10 items, based on lift.  Sorting in desc order
df_final.sort_values('lift', ascending=False).head(10)

lhs                                  rhs  support  \
58          olive oil,    mineral water, whole wheat pasta,      0.004   
6       fromage blanc,                            honey,  ,      0.003   
49        ground beef,             spaghetti, tomato sauce,      0.003   
1             chicken,                      light cream,  ,      0.005   
3            escalope,                            pasta,  ,      0.006   
28        ground beef,         french fries, herb & pepper,      0.003   
11              pasta,                           shrimp,  ,      0.005   
23        ground beef,            chocolate, herb & pepper,      0.004   
69  frozen vegetables,       mineral water, shrimp, chocolate    0.003   
10          olive oil,                whole wheat pasta,  ,      0.008   

    confidance  lift  
58       0.059 6.116  
6        0.245 5.164  
49       0.031 4.981  
1        0.076 4.844  
3        0.074 4.701  
28       0.033 4.697  
11       0.322 4.507  
23       0.041 4.490  
69       0.034 4.417  
10       0.121 4.122

Other way of doing Apriori in Python
--

<h3>In the <u>Second way</u> we would use <b>mlxtend.frequent_patterns</b> package.</h3>

<font color='red'><b>Why we doing it in this way</b></font> - 

1.  **Limitation of first approach** was need to convert data in a list fomat. In real life a store has many thousands of transactations hence it is computationally expensive.

2. `Apyori package is outdated`. 

3. Results are coming in less friendly format so their is a need for pre-processing.

4. Instead lets use **`mlxtend`** package. It generate frequent itemset and association rules both. 

In [41]:
'''
load apriori and association modules from mlxtend.frequent_patterns 
Used different dataset because mlxtend need data in below format. 

    transaction_name    apple banana grapes
transaction  1            0     1      1
             2            1     0      1  
             3            1     0      0
             4            0     1      0
             
 we could have used above data as well but need to perform operation to bring in this format instead of that used seperate data only.            
'''

import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

df1 = pd.read_csv('../data/data-2.csv', encoding="ISO-8859-1")
df1.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26      2.550   17850.000  United Kingdom  
1  12/1/2010 8:26      3.390   17850.000  United Kingdom  
2  12/1/2010 8:26      2.750   17850.000  United Kingdom  
3  12/1/2010 8:26      3.390   17850.000  United Kingdom  
4  12/1/2010 8:26      3.390   17850.000  United Kingdom

In [42]:
# data has many country choose any one for check.
df1.Country.value_counts().head(5)

Country
United Kingdom    495478
Germany             9495
France              8557
EIRE                8196
Spain               2533
Name: count, dtype: int64

In [43]:
# using only France country data for now;  can check for other as well..
df1 = df1[df1.Country == 'France']

In [44]:
# some spaces are there in description; need to remove else later operation it will create problem..
df1['Description'] = df1['Description'].str.strip()

In [45]:
#some of transaction quantity is negative which can not be possible remove them.
df1 = df1[df1.Quantity >0]

In [46]:
#df1[df1.Country == 'France'].head(10)
df1.head(10)

InvoiceNo StockCode                         Description  Quantity  \
26    536370     22728           ALARM CLOCK BAKELIKE PINK        24   
27    536370     22727            ALARM CLOCK BAKELIKE RED        24   
28    536370     22726          ALARM CLOCK BAKELIKE GREEN        12   
29    536370     21724     PANDA AND BUNNIES STICKER SHEET        12   
30    536370     21883                     STARS GIFT TAPE        24   
31    536370     10002          INFLATABLE POLITICAL GLOBE        48   
32    536370     21791   VINTAGE HEADS AND TAILS CARD GAME        24   
33    536370     21035      SET/2 RED RETROSPOT TEA TOWELS        18   
34    536370     22326  ROUND SNACK BOXES SET OF4 WOODLAND        24   
35    536370     22629                  SPACEBOY LUNCH BOX        24   

       InvoiceDate  UnitPrice  CustomerID Country  
26  12/1/2010 8:45      3.750   12583.000  France  
27  12/1/2010 8:45      3.750   12583.000  France  
28  12/1/2010 8:45      3.750   12583.000  France  
29  12/1/2010 8:45      0.850   12583.000  France  
30  12/1/2010 8:45      0.650   12583.000  France  
31  12/1/2010 8:45      0.850   12583.000  France  
32  12/1/2010 8:45      1.250   12583.000  France  
33  12/1/2010 8:45      2.950   12583.000  France  
34  12/1/2010 8:45      2.950   12583.000  France  
35  12/1/2010 8:45      1.950   12583.000  France

In [47]:
# convert data in format which is required 
# converting using pivot table and Quantity sum as values. fill 0 if any nan values

basket = pd.pivot_table(data=df1,index='InvoiceNo',columns='Description',values='Quantity', aggfunc='sum',fill_value=0)

In [48]:
basket.head()

Description  10 COLOUR SPACEBOY PEN  12 COLOURED PARTY BALLOONS  \
InvoiceNo                                                         
536370                            0                           0   
536852                            0                           0   
536974                            0                           0   
537065                            0                           0   
537463                            0                           0   

Description  12 EGG HOUSE PAINTED WOOD  12 MESSAGE CARDS WITH ENVELOPES  \
InvoiceNo                                                                 
536370                               0                                0   
536852                               0                                0   
536974                               0                                0   
537065                               0                                0   
537463                               0                                0   

Description  12 PENCIL SMALL TUBE WOODLAND  \
InvoiceNo                                    
536370                                   0   
536852                                   0   
536974                                   0   
537065                                   0   
537463                                   0   

Description  12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS SMALL TUBE SKULL  \
InvoiceNo                                                                       
536370                                         0                            0   
536852                                         0                            0   
536974                                         0                            0   
537065                                         0                            0   
537463                                         0                            0   

Description  12 PENCILS TALL TUBE POSY  12 PENCILS TALL TUBE RED RETROSPOT  \
InvoiceNo                                                                    
536370                               0                                   0   
536852                               0                                   0   
536974                               0                                   0   
537065                               0                                   0   
537463                               0                                   0   

Description  12 PENCILS TALL TUBE WOODLAND  \
InvoiceNo                                    
536370                                   0   
536852                                   0   
536974                                   0   
537065                                   0   
537463                                   0   

Description  15CM CHRISTMAS GLASS BALL 20 LIGHTS  \
InvoiceNo                                          
536370                                         0   
536852                                         0   
536974                                         0   
537065                                         0   
537463                                         0   

Description  16 PIECE CUTLERY SET PANTRY DESIGN  \
InvoiceNo                                         
536370                                        0   
536852                                        0   
536974                                        0   
537065                                        0   
537463                                        0   

Description  18PC WOODEN CUTLERY SET DISPOSABLE  20 DOLLY PEGS RETROSPOT  \
InvoiceNo                                                                  
536370                                        0                        0   
536852                                        0                        0   
536974                                        0                        0   
537065                                        0                        0   
537463                                        0                 

In [49]:
#this to check correctness after binning it to 1 ..
basket['10 COLOUR SPACEBOY PEN'].head(10)

InvoiceNo
536370     0
536852     0
536974     0
537065     0
537463     0
537468    24
537693     0
537897     0
537967     0
538008     0
Name: 10 COLOUR SPACEBOY PEN, dtype: int64

In [50]:
# we dont need quantity sum 
# we need either has taken or not 
# so if user has taken that item mark as 1 else mark as 0.

def convert_into_binary(x):
    if x > 0:
        return 1
    else:
        return 0

In [51]:
basket_sets = basket.applymap(convert_into_binary)

In [52]:
# check : has quantity now converted to 1 or 0.
basket_sets['10 COLOUR SPACEBOY PEN'].head(10)

InvoiceNo
536370    0
536852    0
536974    0
537065    0
537463    0
537468    1
537693    0
537897    0
537967    0
538008    0
Name: 10 COLOUR SPACEBOY PEN, dtype: int64

In [53]:
# remove postage item as it is just a seal which almost all transaction contains. 
print(basket_sets['POSTAGE'].head())

basket_sets.drop(columns=['POSTAGE'],inplace=True)

InvoiceNo
536370    1
536852    1
536974    1
537065    1
537463    1
Name: POSTAGE, dtype: int64


In [54]:
# call apriori function and pass minimum support here we are passing 7%. 
# means 7 times in total number of transaction the item should be present.
frequent_itemsets = apriori(basket_sets, min_support=0.07, use_colnames=True)

In [55]:
#it will generate frequent itemsets 
frequent_itemsets

support                                           itemsets
0     0.071                      (4 TRADITIONAL SPINNING TOPS)
1     0.097                       (ALARM CLOCK BAKELIKE GREEN)
2     0.102                        (ALARM CLOCK BAKELIKE PINK)
3     0.094                         (ALARM CLOCK BAKELIKE RED)
4     0.082                     (BAKING SET 9 PIECE RETROSPOT)
5     0.071                     (CHILDRENS CUTLERY DOLLY GIRL)
6     0.099                             (DOLLY GIRL LUNCH BOX)
7     0.097                          (JUMBO BAG RED RETROSPOT)
8     0.077                       (JUMBO BAG WOODLAND ANIMALS)
9     0.125                           (LUNCH BAG APPLE DESIGN)
10    0.084                      (LUNCH BAG DOLLY GIRL DESIGN)
11    0.153                          (LUNCH BAG RED RETROSPOT)
12    0.120                        (LUNCH BAG SPACEBOY DESIGN)
13    0.117                               (LUNCH BAG WOODLAND)
14    0.143                 (LUNCH BOX WITH CUTLERY RETROSPOT)
15    0.105                           (MINI PAINT SET VINTAGE)
16    0.102                  (PACK OF 72 RETROSPOT CAKE CASES)
17    0.082                          (PAPER BUNTING RETROSPOT)
18    0.168                    (PLASTERS IN TIN CIRCUS PARADE)
19    0.138                         (PLASTERS IN TIN SPACEBOY)
20    0.082                        (PLASTERS IN TIN STRONGMAN)
21    0.171                 (PLASTERS IN TIN WOODLAND ANIMALS)
22    0.189                               (RABBIT NIGHT LIGHT)
23    0.097                      (RED RETROSPOT CHARLOTTE BAG)
24    0.138                         (RED RETROSPOT MINI CASES)
25    0.071                         (RED RETROSPOT PICNIC BAG)
26    0.181                    (RED TOADSTOOL LED NIGHT LIGHT)
27    0.125                         (REGENCY CAKESTAND 3 TIER)
28    0.087                  (RETROSPOT TEA SET CERAMIC 11 PC)
29    0.107                (ROUND SNACK BOXES SET OF 4 FRUITS)
30    0.158               (ROUND SNACK BOXES SET OF4 WOODLAND)
31    0.077                (SET/10 RED POLKADOT PARTY CANDLES)
32    0.133               (SET/20 RED RETROSPOT PAPER NAPKINS)
33    0.138                      (SET/6 RED SPOTTY PAPER CUPS)
34    0.128                    (SET/6 RED SPOTTY PAPER PLATES)
35    0.071                           (SPACEBOY BIRTHDAY CARD)
36    0.125                               (SPACEBOY LUNCH BOX)
37    0.122                (STRAWBERRY LUNCH BOX WITH CUTLERY)
38    0.094                          (TEA PARTY BIRTHDAY CARD)
39    0.074                           (WOODLAND CHARLOTTE BAG)
40    0.074  (ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKEL...
41    0.079  (ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKEL...
42    0.074  (ALARM CLOCK BAKELIKE PINK, ALARM CLOCK BAKELI...
43    0.071         (SPACEBOY LUNCH BOX, DOLLY GIRL LUNCH BOX)
44    0.089  (PLASTERS IN TIN SPACEBOY, PLASTERS IN TIN CIR...
45    0.102  (PLASTERS IN TIN WOODLAND ANIMALS, PLASTERS IN...
46    0.105  (PLASTERS IN TIN SPACEBOY, PLASTERS IN TIN WOO...
47    0.102  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...
48    0.102  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...
49    0.122  (SET/6 RED SPOTTY PAPER PLATES, SET/6 RED SPOT...
50    0.099  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...

Generating frequent itemsets from a list of items - 
--

First step in generation of association rules is to get all the frequent itemsets on which binary partitions can be performed to get the antecedent and the consequent. 


<img src='https://miro.medium.com/max/576/1*NoUFBxsokdnw3wrYyVXkgw.png' />


`For example`, if there are 6 items {Bread, Butter, Egg, Milk, Notebook, Toothbrush} on all the transactions combined, itemsets will look like {Bread}, {Butter}, {Bread, Notebook}, {Milk, Toothbrush}, {Milk, Egg, Vegetables} etc. Size of an itemset can vary from one to the total number of items that we have. Now, we seek only frequent itemsets from this and not all so as to put a check on the number of total itemsets generated.

<img src='https://miro.medium.com/max/256/1*gZujUCrD9nanRr-ETBFtbg.png' />

<br>

Frequent itemsets are the ones which occur at least a minimum number of times in the transactions. Technically, these are the itemsets for which support value (fraction of transactions containing the itemset) is above a **minimum threshold — min_support. We have kept min_support=0.07 in our above code**.

So, {Bread, Notebook} might not be a frequent itemset if it occurs only 2 times out of 100 transactions and (2/100) = 0.02 falls below the value of min_support.

In [56]:
# We would apply association rules on frequent itemset. 
# here we are setting based on lift and keeping minimum lift as 1

rules_mlxtend = association_rules(frequent_itemsets, metric="lift", min_threshold=1)
rules_mlxtend.head()

antecedents                   consequents  \
0  (ALARM CLOCK BAKELIKE GREEN)   (ALARM CLOCK BAKELIKE PINK)   
1   (ALARM CLOCK BAKELIKE PINK)  (ALARM CLOCK BAKELIKE GREEN)   
2  (ALARM CLOCK BAKELIKE GREEN)    (ALARM CLOCK BAKELIKE RED)   
3    (ALARM CLOCK BAKELIKE RED)  (ALARM CLOCK BAKELIKE GREEN)   
4   (ALARM CLOCK BAKELIKE PINK)    (ALARM CLOCK BAKELIKE RED)   

   antecedent support  consequent support  support  confidence  lift  \
0               0.097               0.102    0.074       0.763 7.479   
1               0.102               0.097    0.074       0.725 7.479   
2               0.097               0.094    0.079       0.816 8.643   
3               0.094               0.097    0.079       0.838 8.643   
4               0.102               0.094    0.074       0.725 7.681   

   leverage  conviction  zhangs_metric  
0     0.064       3.791          0.959  
1     0.064       3.284          0.965  
2     0.070       4.916          0.979  
3     0.070       5.569          0.976  
4     0.064       3.293          0.969

Generating all possible rules from the frequent itemsets
--

Once the frequent itemsets are generated, identifying rules out of them is comparatively less taxing. Rules are formed by binary partition of each itemset. If {Bread,Egg,Milk,Butter} is the frequent itemset, candidate rules will look like:

(Egg, Milk, Butter → Bread), (Bread, Milk, Butter → Egg), (Bread, Egg → Milk, Butter), (Egg, Milk → Bread, Butter), (Butter→ Bread, Egg, Milk)

From a list of all possible candidate rules, we aim to identify rules that fall above a minimum threshold level (like min_confidence or min_lift).



In [57]:
# rules_mlxtend.rename(columns={'antecedents':'lhs','consequents':'rhs'})

# as based business use case we can sort based on confidance and lift.
rules_mlxtend[ (rules_mlxtend['lift'] >= 4) & (rules_mlxtend['confidence'] >= 0.8) ].head()

antecedents                           consequents  \
2      (ALARM CLOCK BAKELIKE GREEN)            (ALARM CLOCK BAKELIKE RED)   
3        (ALARM CLOCK BAKELIKE RED)          (ALARM CLOCK BAKELIKE GREEN)   
17  (SET/6 RED SPOTTY PAPER PLATES)  (SET/20 RED RETROSPOT PAPER NAPKINS)   
18  (SET/6 RED SPOTTY PAPER PLATES)         (SET/6 RED SPOTTY PAPER CUPS)   
19    (SET/6 RED SPOTTY PAPER CUPS)       (SET/6 RED SPOTTY PAPER PLATES)   

    antecedent support  consequent support  support  confidence  lift  \
2                0.097               0.094    0.079       0.816 8.643   
3                0.094               0.097    0.079       0.838 8.643   
17               0.128               0.133    0.102       0.800 6.031   
18               0.128               0.138    0.122       0.960 6.969   
19               0.138               0.128    0.122       0.889 6.969   

    leverage  conviction  zhangs_metric  
2      0.070       4.916          0.979  
3      0.070       5.569          0.976  
17     0.085       4.337          0.956  
18     0.105      21.556          0.982  
19     0.105       7.852          0.993

How to read the above data ?
--

> antecedents and consequents  -> The IF component of an association rule is known as the antecedent. The THEN component is known as the consequent. The antecedent and the consequent are disjoint; they have no items in common.

> Example of antecedents and consequents -> Consider antecedents as X and consequents as Y.  

<img src='https://miro.medium.com/max/576/1*NoUFBxsokdnw3wrYyVXkgw.png' />

> antecedent support -> This measure gives an idea of how frequent antecedent is in all the transactions. Like (ALARM CLOCK BAKELIKE RED) is present in 9.4% of the transactions. 	

> consequent support -> This measure gives an idea of how frequent consequent is in all the transactions. Like (ALARM CLOCK BAKELIKE GREEN) is present in 9.6% of the transactions. 

> support -> This measure gives an idea of how frequent `ItemSet` is in all the transactions. Like {ALARM CLOCK BAKELIKE RED,ALARM CLOCK BAKELIKE GREEN} is present in 7.9% of the transactions. 

> confidance -> This measure defines the likeliness of occurrence of consequent on the cart given that the cart already has the antecedents. So {ALARM CLOCK BAKELIKE RED} -> {ALARM CLOCK BAKELIKE GREEN} as a confidence of 83%. In simple words their is an 83% chance of finding {ALARM CLOCK BAKELIKE GREEN} , if the cart contains {ALARM CLOCK BAKELIKE RED}.

> lift -> This measure defines the likeliness of occurrence of consequent on the cart given that the cart already has the antecedent, but controlling the popularity of consequent. So lift of {ALARM CLOCK BAKELIKE GREEN} w.r.t {ALARM CLOCK BAKELIKE RED} is 8.64. Which is quite good. Any lift value > 1 implies that the Association rule is worth considering.  [ Reference : https://en.wikipedia.org/wiki/Lift_(data_mining)]

> leverage -> leverage(X -> Y) = P(X and Y) - (P(X)P(Y))

*Leverage measures the difference of X and Y appearing together in the data set and what would be expected if X and Y where statistically dependent. The rational in a sales setting is to find out how many more units (items X and Y together) are sold than expected from the independent sells.*  [Reference : https://michael.hahsler.net/research/recommender/associationrules.html]

> conviction -> Ignore this parameter.  Not much of use in most situations.
